server/src/main/java/org/opensearch/index/engine/EngineConfig.java

```diff
@@ -148,13 +151,42 @@ public Supplier<RetentionLeases> retentionLeasesSupplier() {
      * Compression Level gives a trade-off between compression ratio and speed. The higher compression level results in higher compression ratio but slower compression and decompression speeds.
      * This setting is <b>not</b> realtime updateable.
      */
-    public static final Setting<Integer> INDEX_CODEC_COMPRESSION_LEVEL_SETTING = Setting.intSetting(
+
+    public static final Setting<Integer> INDEX_CODEC_COMPRESSION_LEVEL_SETTING = new Setting<>(
         "index.codec.compression_level",
-        3,
-        1,
-        6,
+        Integer.toString(3),
+        new Setting.IntegerParser(1, 6, "index.codec.compression_level", false),
         Property.IndexScope
-    );
+    ) {
+        @Override
+        public Set<SettingDependency> getSettingsDependencies(String key) {
+            return Set.of(new SettingDependency() {
+                @Override
+                public Setting<String> getSetting() {
+                    return INDEX_CODEC_SETTING;
+                }
+
+                @Override
+                public void validate(String key, Object value, Object dependency) {
+                    if (!(dependency instanceof String)) {
+                        throw new IllegalArgumentException("Codec should be of string type.");
+                    }
+                    doValidateCodecSettings((String) dependency);
```

Comment:
Ok, @sarthakaggarwal97 , I think we have a solution here, we have to follow the same principle as `INDEX_CODEC_SETTING`: in general, codecs are dynamic and could be supplied by plugins (along with the codec service itself). But for compression level (and possibly all other different settings), we could have a marker interface:

```
public interface CodecSettings {
    boolean supports(Setting<?> setting);
}
``` 

Than validation becomes easy:

```
@Override
public void validate(String key, Object value, Object dependency) {
    switch (value) {
          case "best_compression":
          case "lucene_default":
          case "default":
               break;

           default:
                if (Codec.availableCodecs().contains(s)) {
                          Codec codec = Codec.forName(s);
                          if(codec instanceof CodecSettings && ((CodecSettings)codec).supports(INDEX_CODEC_COMPRESSION_LEVEL_SETTING)) {
                              return;
                          }
                }
throw new IllegalArgumentException(
                "Compression level cannot be set for the "
                    + codec
                    + " codec. Compression level settings is only applicable for zstd and zstd_no_dict codecs."
            );

}

Reply:
thanks @reta for recommending the solution, I've refactored the changes, the build seems to failing at the unrelated bwc check, checking this